components/engine_traits/src/engine.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::fmt::Debug;
+use std::path::Path;
+
+use crate::*;
+
+pub trait Snapshot: 'static + Peekable + Send + Sync + Debug {
+    fn cf_names(&self) -> Vec<&str>;
+}
+
+pub trait WriteBatch: Mutable {
+    fn data_size(&self) -> usize;
+    fn count(&self) -> usize;
+    fn is_empty(&self) -> bool;
+    fn clear(&self);
+
+    fn set_save_point(&mut self);
+    fn pop_save_point(&mut self) -> Result<()>;
+    fn rollback_to_save_point(&mut self) -> Result<()>;
+}
+
+pub trait KvEngine: Peekable + Mutable + Iterable + Send + Sync + Clone {
+    type Snap: Snapshot;
+    type Batch: WriteBatch;
+
+    fn write_opt(&self, opts: &WriteOptions, wb: &Self::Batch) -> Result<()>;
+    fn write(&self, wb: &Self::Batch) -> Result<()> {
+        self.write_opt(&WriteOptions::default(), wb)
+    }
+    fn write_batch(&self, cap: usize) -> Self::Batch;
+    fn snapshot(&self) -> Self::Snap;
+    fn sync(&self) -> Result<()>;
+    fn cf_names(&self) -> Vec<&str>;
+    fn delete_all_in_range(
+        &self,
+        t: DeleteRangeType,
+        start_key: &[u8],
+        end_key: &[u8],
+    ) -> Result<()> {
+        if start_key >= end_key {
+            return Ok(());
+        }
+        for cf in self.cf_names() {
+            self.delete_all_in_range_cf(t, cf, start_key, end_key)?;
+        }
+        Ok(())
+    }
+    fn delete_all_in_range_cf(
+        &self,
+        t: DeleteRangeType,
+        cf: &str,
+        start_key: &[u8],
+        end_key: &[u8],
+    ) -> Result<()>;
+    fn ingest_external_file_cf(
+        &self,
+        opts: &IngestExternalFileOptions,
```

Comment:
How about removing it? TiKV always uses the move option provided by RocksDB.

Reply:
Do we need to use other options in the future?

Reply:
Not sure, but most likely not.